File: internal/protocol/operations.go
package protocol

import "github.com/john6604/yata-collaborative-editor/internal/identifier"

type InsertOperation struct {
	NewID    identifier.ID
	OriginID identifier.ID
	RightID  identifier.ID
	Content  byte
}

type DeleteOperation struct {
	TargetID identifier.ID
}

func NewInsertOperation(newID identifier.ID, originID identifier.ID, rightID identifier.ID, content byte) *InsertOperation {
	insert := InsertOperation{NewID: newID}
	insert.OriginID = originID
	insert.RightID = rightID
	insert.Content = content
	return &insert
}

func NewDeleteOperation(targetID identifier.ID) *DeleteOperation {
	delete := DeleteOperation{TargetID: targetID}
	return &delete
}

File: frontend/src/crdt/operations.js
export class InsertOperation {
    constructor(newID, originID, rightID, content) {
        this.new_id = newID;
        this.origin_id = originID;
        this.right_id = rightID;
        this.content = content;
    }
}

export class DeleteOperation {
    constructor(targetID) {
        this.target_id = targetID;
    }
}

export class Sync1Operation {
    constructor(vectorState, deleteSet) {
        this.vector_state = vectorState;
        this.delete_set = deleteSet;
    }
}

export class Sync2Operation {
    constructor(delta) {
        this.delta = delta;
    }
}

export class SnapshotOperation {
    constructor(delta) {
        this.delta = delta;
    }
}

export class PresenceOperation {
    constructor(users) {
        this.users = users;
    }
}